File: src/main/java/com/vaadin/demo/component/combobox/ComboBoxValidation.java
package com.vaadin.demo.component.combobox;

import com.vaadin.flow.component.combobox.ComboBox;
import com.vaadin.flow.component.orderedlayout.HorizontalLayout;
import com.vaadin.flow.router.Route;
import com.vaadin.demo.DemoExporter; // hidden-source-line

@Route("combo-box-validation")
public class ComboBoxValidation extends HorizontalLayout {

    public ComboBoxValidation() {
        setPadding(false);

        // tag::snippet[]
        ComboBox<String> field = new ComboBox<>();
        field.setRequiredIndicatorVisible(true);
        field.setAllowedCharPattern("[A-Z]");
        // end::snippet[]
        field.setLabel("Country code");
        field.setHelperText("2-letter uppercase ISO country code");
        field.setItems("DE", "FI", "US");
        field.setAllowCustomValue(true);

        add(field);
    }

    public static class Exporter extends DemoExporter<ComboBoxValidation> { // hidden-source-line
    } // hidden-source-line
}

File: frontend/demo/component/combobox/combo-box-validation.ts
import 'Frontend/demo/init'; // hidden-source-line
import '@vaadin/combo-box';
import { html, LitElement } from 'lit';
import { customElement } from 'lit/decorators.js';
import { applyTheme } from 'Frontend/demo/theme';

@customElement('combo-box-validation')
export class Example extends LitElement {
  protected override createRenderRoot() {
    const root = super.createRenderRoot();
    applyTheme(root);
    return root;
  }

  protected override render() {
    return html`
      <!-- tag::snippet[] -->
      <vaadin-combo-box
        required
        allowed-char-pattern="[A-Z]"
        label="Country code"
        helper-text="2-letter uppercase ISO country code"
        error-message="Field is required"
        allow-custom-value
        .items="${['DE', 'FI', 'US']}"
      ></vaadin-combo-box>
      <!-- end::snippet[] -->
    `;
  }
}
